core/types/calls.cc

```diff
@@ -1461,36 +1461,54 @@ class T_attached_class : public IntrinsicMethod {
+namespace {
+
+enum class MustBehavior {
+    T,
+    Magic,
+};
+
+void implementMust(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res, MustBehavior behavior) {
+    if (args.args.empty()) {
+        return;
+    }
+    const auto loc = args.callLoc();
+    if (!args.args[0]->type.isFullyDefined()) {
+        if (behavior == MustBehavior::Magic) {
+            /* Nothing */
+        } else if (auto e = gs.beginError(loc, errors::Infer::BareTypeUsage)) {
+            e.setHeader("T.must() applied to incomplete type `{}`", args.args[0]->type.show(gs));
         }
-        const auto loc = args.callLoc();
-        if (!args.args[0]->type.isFullyDefined()) {
-            if (auto e = gs.beginError(loc, errors::Infer::BareTypeUsage)) {
-                e.setHeader("T.must() applied to incomplete type `{}`", args.args[0]->type.show(gs));
+        return;
+    }
+    auto ret = Types::approximateSubtract(gs, args.args[0]->type, Types::nilClass());
+    if (ret == args.args[0]->type) {
+        if (behavior == MustBehavior::Magic) {
+            /* Nothing */
+        } else if (auto e = gs.beginError(loc, errors::Infer::InvalidCast)) {
+            if (args.args[0]->type.isUntyped()) {
+                e.setHeader("`{}` called on `{}`, which is redundant", "T.must", args.args[0]->type.show(gs));
+            } else {
+                e.setHeader("`{}` called on `{}`, which is never `{}`", "T.must", args.args[0]->type.show(gs), "nil");
             }
-            return;
-        }
-        auto ret = Types::approximateSubtract(gs, args.args[0]->type, Types::nilClass());
-        if (ret == args.args[0]->type) {
-            if (auto e = gs.beginError(loc, errors::Infer::InvalidCast)) {
-                if (args.args[0]->type.isUntyped()) {
-                    e.setHeader("`{}` called on `{}`, which is redundant", "T.must", args.args[0]->type.show(gs));
-                } else {
-                    e.setHeader("`{}` called on `{}`, which is never `{}`", "T.must", args.args[0]->type.show(gs),
-                                "nil");
-                }
-                e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
-                const auto locWithoutTMust = loc.adjust(gs, 7, -1);
-                if (loc.exists() && locWithoutTMust.exists()) {
-                    e.replaceWith("Remove `T.must`", loc, "{}", locWithoutTMust.source(gs).value());
-                }
+            e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
+            const auto locWithoutTMust = loc.adjust(gs, 7, -1);
+            if (loc.exists() && locWithoutTMust.exists()) {
+                e.replaceWith("Remove `T.must`", loc, "{}", locWithoutTMust.source(gs).value());
             }
         }
-        res.returnType = move(ret);
+    }
+    if (behavior == MustBehavior::Magic) {
+        ret = Types::approximateSubtract(gs, ret, Types::falseClass());
+    }
+    res.returnType = move(ret);
```

Comment:
> It's not clear to me that res.returnType will always be appropriately defined after we return from this function, hence doing it here.

I'm not sure how it wouldn't be? I'm suggesting this here:

```suggestion
    res.returnType = move(ret);
```

and this in `Magic_mustBeTruthy` below:

```cpp
        implementMust(gs, args, res, MustBehavior::Magic);
        auto resultType = res.returnType;
        res.returnType = Types::approximateSubtract(gs, resultType, Types::falseClass());
```

Reply:
Sure, but you can return from that function without setting `res.returnType`, no?  Or is it true that `res.returnType` is always set to *something*, and intrinsic functions are just refining the type in some way?

Reply:
> Sure, but you can return from that function without setting `res.returnType`, no? Or is it true that `res.returnType` is always set to _something_, and intrinsic functions are just refining the type in some way?

Yes, if you don't set the `DispatchResult::returnType`, it will just stay as what it was default initialized to (`nullptr`) and fall through to being set using the `resultType` if a signature is present:

https://github.com/sorbet/sorbet/blob/db1373aa424b5fe1a02e1e71d4032cf0ebd6eeac/core/types/calls.cc#L1176-L1186

or defaulted to `T.untyped` if no signature was present:

https://github.com/sorbet/sorbet/blob/db1373aa424b5fe1a02e1e71d4032cf0ebd6eeac/core/types/calls.cc#L1210-L1211

**BUT** in neither case here are we not setting it. In `implementMust` we'd be setting to `ret` (unconditionally) and only in `Magic_mustBeTruthy` would we be mutating it to drop subtypes of `FalseClass`.

Reply:
Thanks for the explanation.  I still think it makes more sense to have the subtraction of `FalseClass` in `implementMust` itself, but :man_shrugging: .